### ISSUE TITLE
Strange reference cycle (?) introduced with shim frames

### ISSUE THREAD

**Issue by A5rocks** [python/cpython#100964]
# Bug report

https://github.com/python-trio/trio/pull/2514#issuecomment-1378120405

```py
import weakref
import trio
import gc

async def f():
    async with trio.open_nursery() as nursery:
        nursery.cancel_scope.cancel()

    h = weakref.ref(nursery)
    del nursery
    await trio.lowlevel.checkpoint()
    print(h(), gc.get_referrers(h()) if h() is not None else [])

trio.run(f)
```

(I haven't figured out how to reproduce without the `import trio`)

I bisected this to https://github.com/python/cpython/commit/1e197e63e21f77b102ff2601a549dda4b6439455

Previously:
```
None []
```

Now:
```
<trio.Nursery object at 0x7f3f5585d4f0> [NurseryManager(strict_exception_groups=False), <cell at 0x7f3f5585d540: Nursery object at 0x7f3f5585d4f0>, <frame at 0x7f3f55e220c0, file '/workspaces/trio/trio/_core/_run.py', line 1005, code _nested_child_finished>]
```

I'm not certain this should be a bug, but raising because I'm confused as to the cause.

### Linked PRs
* gh-107563
* gh-141112
* gh-142902
* gh-142904

**Comment by brandtbucher** [comment]
I believe that commit slightly changed how/when generators clear their frames upon raising or returning. I forget exactly how, though... I think it might be that the generator used to link, unlink, and clear its own frame in `gen_send_ex2` but now all that work happens in the VM? Or something.

Seems like it could definitely be related.

**Comment by A5rocks** [comment]
Yup, this issue is just about new cycles being created where there were none before.

The context for the testcase is a regression test for https://github.com/python-trio/trio/issues/1770 (which kinda explains why we care about this in the first place).

**Comment by iritkatriel** [comment]
You can add gc.collect() to the test.

**Comment by A5rocks** [comment]
Sorry, I think you might be understanding me. In order to minimize stalls, in trio we've put in some extra work to remove reference cycles (and to make GC run less often). This new shim frames mechanism adds some reference cycles which probably shouldn't be added -- I haven't looked at CPython internals to know why, but the news message says "The extra frame should be invisible to all Python and most C extensions." Adding `gc.collect()` "fixes" this but that's not really the ideal solution for us; users probably shouldn't have to think about GC.

**Comment by iritkatriel** [comment]
Ok, so I'll label is performance instead of bug.

**Comment by markshannon** [comment]
Note to self:
It is possible that we need to clear `gi_exc_state.exc_value` in `clear_gen_frame()`

**Comment by iritkatriel** [comment]
Is this our cycle?

```
(Pdb) p ref
<built-in method send of coroutine object at 0x104bf6c30>
(Pdb) p gc.get_referrers(ref)[1]
<coroutine object f at 0x104bf6c30>
(Pdb) p gc.get_referrers(gc.get_referrers(ref)[1])[1]
<built-in method send of coroutine object at 0x104bf6c30>
```

**Comment by markshannon** [comment]
@A5rocks #100964 may fix this issue. Can you test it for me?
I get an error trying to pip install `trio` with the latest CPython build.

**Comment by A5rocks** [comment]
This was fixed in trio but it turns out that fix was because https://github.com/python-trio/trio/commit/a3d9d2536a4fefd9bfce2b66fa6a557304c54c81

... I'll dedicate some more time to hopefully pinning this down because even though this isn't very important it still bothers me!

**Comment by A5rocks** [comment]
hi.

i've found a minimal reproduction that only relies on the standard library. this feels so cursed.

```py
import weakref
import asyncio

async def preserve(a):
    try:
        raise Exception("aaa")
    except BaseException as exc:
        e = exc

    try:
        return e
    finally:
        del e


class Dummy:
    def __repr__(self):
        return "ahahahahahah its broken!!"

class PersonManager:
    async def __aenter__(self):
        self._person = Dummy()
        return self._person

    async def __aexit__(self, etype, exc, tb):
        new_exc = await preserve(self._person)
        return True

async def f():
    async with PersonManager() as me:
        pass

    h = weakref.ref(me)
    del me
    await asyncio.sleep(0.01)
    print(h())

asyncio.run(f())
```

sorry for taking so long.

**Comment by brandtbucher** [comment]
Thanks, this is super helpful. Smaller reproducer that prints `True` on 3.11 and and `False` on 3.12:

```py
import weakref
import asyncio

async def preserve():
    try:
        raise Exception()
    except Exception as exc:
        return exc

class PeopleManager:
    pass

def __aexit__(self):
    new_exc = asyncio.run(preserve())

def f():
    manager = PeopleManager()
    __aexit__(manager)
    h = weakref.ref(manager)
    del manager
    print(h() is None)

f()
```

**Comment by brandtbucher** [comment]
New minimal reproducer, using a generator instead of `asyncio`:

```py
import weakref

def g():
    try:
        raise Exception()
    except Exception as e:
        yield e

class C:
    pass

def f(o):
    ref = next(g())

o = C()
f(o)
h = weakref.ref(o)
del o
print(h() is None)
```

The cycle is `ref is ref.__traceback__.tb_frame.f_back.f_locals["ref"]`. This cycle keeps the locals, including `o`, alive.

On 3.11 and earlier, `f_back` is `None` (as it should be, since the exception was caught in the raising frame). On 3.12, there is a full chain of frames when an exception is raised in a generator/coroutine, regardless of where it was caught. That's not great, since we're potentially keeping huge graphs of objects alive every time one of these exceptions is stashed away...

**Comment by brandtbucher** [comment]
New minimal test (prints `True` on 3.11, `False` on 3.12). So yeah, I think the generator isn't getting unlinked for some reason:

```py
def g():
    try:
        raise Exception()
    except Exception as e:
        yield e

x = next(g())
print(x.__traceback__.tb_frame.f_back is None)
```

Note that if I do `gen = g(); x = next(g)`, 3.12 will print `True` again. So I'm guessing it may have something to do with generator finalization, too.
